opencensus/trace/trace_state.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2017, OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+import re
+
+_KEY_FORMAT = r'[a-z][_0-9a-z\-\*\/]{0,255}'
+_VALUE_FORMAT = r'[\x20-\x2b\x2d-\x3c\x3e-\x7e]{1,256}'
+
+_KEY_VALIDATION_RE = re.compile('^' + _KEY_FORMAT + '$')
+_VALUE_VALIDATION_RE = re.compile('^' + _VALUE_FORMAT + '$')
+
+
+class TraceState(OrderedDict):
+    def __setitem__(self, key, value):
+        if not isinstance(key, str):
+            raise ValueError('key must be an instance of str')
+        if not re.match(_KEY_VALIDATION_RE, key):
+            raise ValueError('illegal key provided')
+        if not isinstance(value, str):
+            raise ValueError('value must be an instance of str')
+        if not re.match(_VALUE_VALIDATION_RE, value):
+            raise ValueError('illegal value provided')
+        super(TraceState, self).__setitem__(key, value)
+
+    def append(self, key, value):
```

Comment:
It looks like the Tracestate is mutable and this may cause problems (not sure in python with their threading schema): if possible I would prefer to have the Tracestate immutable because once we create it and put in the the SpanContext we don't want to be changed. Scenarios like: create a span with a tracestate then modify the tracestate should not happen in my opinion.

Reply:
This is contrary to my understanding, I was thinking tracestate should allow changes at any point. Would you elaborate?

For example:

with tracer.span('A'):
    if foo():
        current_span.span_context.tracestate['system_x_tracking_id'] = '123'
        invoke_system_x()
    else:
        current_span.span_context.tracestate['system_y_tracking_id'] = '456'
        invoke_system_y()

Reply:
You should be able to scope `tracestate` properties downstream. Like for hierarchical identifiers in MS and skywalking. But modifying upstream sounds like a bad idea. So I think you both are correct. (note, I haven't looked into PR itself yet)